src/home/homework-blob.c

```diff
@@ -0,0 +1,236 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later
+ *
+ * Copyright © 2024 GNOME Foundation Inc.
+ *      Original Author: Adrian Vovk
+ */
+
+#include "copy.h"
+#include "fileio.h"
+#include "fd-util.h"
+#include "fs-util.h"
+#include "home-util.h"
+#include "homework-blob.h"
+#include "homework.h"
+#include "install-file.h"
+#include "macro.h"
+#include "path-util.h"
+#include "recurse-dir.h"
+#include "rm-rf.h"
+#include "sha256.h"
+#include "string-util.h"
+#include "tmpfile-util.h"
+#include "umask-util.h"
+#include "utf8.h"
+
+#define BLOB_MAX_SIZE (64U*1024U*1024U)
+
+static int copy_one_blob(
+                int src,
+                int dest_dfd,
+                const char *name,
+                uint64_t *total_size,
+                uid_t uid,
+                Hashmap *manifest) {
+        _cleanup_close_ int dest = -EBADF;
+        uint8_t hash[SHA256_DIGEST_SIZE];
+        off_t initial, size;
+        int r;
+
+        if (!hashmap_contains(manifest, name)) {
+                log_warning("File %s in blob directory is missing from manifest. Skipping.", name);
+                return 0;
+        }
+
+        if (!suitable_blob_filename(name)) {
+                log_warning("File %s in blob directory has invalid filename. Skipping.", name);
+                return 0;
+        }
+
```

Comment:
i think it would be wise to fstat() the file here, and then refuse non-regular files via stat_verify_regular().

and maybe even already do the size check early here because we can (but keep the one at the bottom, since people could give use an fd to sysfs/procfs or so where there are files that report a size of zero but which you can read GiB of data from

Reply:
The reason I didn't add this is because we already check this before we call `copy_one_blob`. In the reconcile case, we check in `blob_copy_callback` via statx immediately before we call `copy_one_blob`. In the replace case, we make sure that any FD passed through dbus is a regular file. So `blobs` will only ever contain regular file FDs.

I think what I'll do is add a check in the replace case (where it's not quite so obvious that `blobs` will always contain regular file FDs), and document in the function that FD must point to a regular file.

As for size: if the stat output can lie to us anyway, IDK if it's worthwhile

Reply:
Reworking into `readdir_all` made this make more sense so I did it. Still without the size check; if it's going to be inaccurate sometimes and we're going to check it anyway shortly after and `sha256_fd` won't go on reading for too long, I see no reason to bother. It would be a tad complicated since we'd need to be careful about overflows there too